File: demo2-atp-cli/src/test/java/dev/kanashi/atp/cli/store/TocTouTest.java
package dev.kanashi.atp.cli.store;

import dev.kanashi.atp.cli.model.ExitCode;
import dev.kanashi.atp.cli.model.StoreResult;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * ⭐ 这个类测的是整套设计存在的理由：
 * <b>用户确认的那一份，和最终落库的那一份，必须是同一份。</b>
 *
 * <p>如果 commit 写成"先 SELECT 检查状态和版本、再 UPDATE"，
 * 检查通过之后、UPDATE 执行之前 agent 改一次内容，这里就会静默地提交错的版本。
 * 把状态和版本压进同一条 UPDATE 的 WHERE，窗口才是零。
 */
@DisplayName("TOCTOU：确认之后内容被改过，提交必须失败")
class TocTouTest extends MySqlTestBase {

    @Test
    @DisplayName("preview 拿到 version=1 后被改成 2 → commit(version=1) 报 VERSION_CONFLICT")
    void staleVersionRejected() {
        String caseId = UUID.randomUUID().toString();
        store.draft(caseId, PC_WEB, "购物车结算", "agent-a");

        StoreResult afterFirstEdit = store.update(caseId, 0, completeDraft("购物车结算"));
        int previewedVersion = afterFirstEdit.row().version();   // 用户 preview 看到的就是这个
        assertThat(previewedVersion).isEqualTo(1);

        // 用户点确认之前，agent 又偷偷改了一版
        StoreResult sneaky = store.update(caseId, 1, completeDraft("购物车结算（被改过）"));
        assertThat(sneaky.row().version()).isEqualTo(2);

        StoreResult commit = store.commit(caseId, previewedVersion);

        assertThat(commit.code()).isEqualTo(ExitCode.VERSION_CONFLICT);
        assertThat(commit.message()).contains("被改过");
        assertThat(store.show(caseId).row().status())
                .as("案例必须还停在编写态，不能被提交出去")
                .isEqualTo("AI_DRAFT");
    }

    @Test
    @DisplayName("重新 preview 后用新版本号提交 → 成功")
    void recoverByRepreview() {
        String caseId = UUID.randomUUID().toString();
        store.draft(caseId, PC_WEB, "购物车结算", "agent-a");
        store.update(caseId, 0, completeDraft("购物车结算"));
        store.update(caseId, 1, completeDraft("购物车结算（终稿）"));

        int current = store.show(caseId).row().version();
        StoreResult commit = store.commit(caseId, current);

        assertThat(commit.code()).isEqualTo(ExitCode.OK);
        assertThat(commit.row().status()).isEqualTo("DRAFT");
    }

    @Test
    @DisplayName("并发 update 撞版本 → 后到的被 CAS 挡下")
    void concurrentUpdateLosesOnCas() {
        String caseId = UUID.randomUUID().toString();
        store.draft(caseId, PC_WEB, "登录成功", "agent-a");

        StoreResult win = store.update(caseId, 0, completeDraft("A 写的"));
        StoreResult lose = store.update(caseId, 0, completeDraft("B 写的"));

        assertThat(win.code()).isEqualTo(ExitCode.OK);
        assertThat(lose.code()).isEqualTo(ExitCode.VERSION_CONFLICT);
    }
}

File: demo2-atp-cli/internal/store/toctou_test.go
package store_test

import (
	"testing"

	"github.com/Kanash1i/atp-ai-demos/atp-cli/internal/model"
	"github.com/google/uuid"
)

// 这一组测的是整套设计存在的理由：
// 用户确认的那一份，和最终落库的那一份，必须是同一份。
//
// 如果 commit 写成"先 SELECT 检查状态和版本、再 UPDATE"，
// 检查通过之后、UPDATE 执行之前 agent 改一次内容，这里就会静默地提交错的版本。
// 把状态和版本压进同一条 UPDATE 的 WHERE，窗口才是零。

func TestTocTou_StaleVersionRejected(t *testing.T) {
	ctx, s, _ := newStore(t)
	id := uuid.NewString()
	mustDraft(t, ctx, s, id, "购物车结算")

	afterFirst := s.Update(ctx, id, 0, completeDraft("购物车结算", 2))
	previewed := afterFirst.Row.Version // 用户 preview 看到的就是这个
	if previewed != 1 {
		t.Fatalf("首次 update 后 version 应为 1，实际 %d", previewed)
	}

	// 用户点确认之前，agent 又偷偷改了一版
	sneaky := s.Update(ctx, id, 1, completeDraft("购物车结算（被改过）", 2))
	if sneaky.Row.Version != 2 {
		t.Fatalf("二次 update 后 version 应为 2，实际 %d", sneaky.Row.Version)
	}

	r := s.Commit(ctx, id, previewed)

	if r.Code != model.VersionConflict {
		t.Fatalf("拿过期 version 提交必须被拒，实际 %s: %s", r.Code, r.Message)
	}
	if got := s.Show(ctx, id).Row.Status; got != model.StatusAIDraft {
		t.Fatalf("案例必须还停在编写态，实际 %s", got)
	}
}

func TestTocTou_RecoverByRepreview(t *testing.T) {
	ctx, s, _ := newStore(t)
	id := uuid.NewString()
	mustDraft(t, ctx, s, id, "购物车结算")
	s.Update(ctx, id, 0, completeDraft("购物车结算", 2))
	s.Update(ctx, id, 1, completeDraft("购物车结算（终稿）", 2))

	current := s.Show(ctx, id).Row.Version
	if r := s.Commit(ctx, id, current); r.Code != model.OK {
		t.Fatalf("重新 preview 后用新版本号提交应成功，实际 %s: %s", r.Code, r.Message)
	}
}

func TestTocTou_ConcurrentUpdateLosesOnCAS(t *testing.T) {
	ctx, s, _ := newStore(t)
	id := uuid.NewString()
	mustDraft(t, ctx, s, id, "登录成功")

	win := s.Update(ctx, id, 0, completeDraft("A 写的", 1))
	lose := s.Update(ctx, id, 0, completeDraft("B 写的", 1))

	if win.Code != model.OK {
		t.Fatalf("先到的应成功，实际 %s", win.Code)
	}
	if lose.Code != model.VersionConflict {
		t.Fatalf("后到的应被 CAS 挡下，实际 %s", lose.Code)
	}
}

// TestCAS_SerialAdvanceAlwaysSucceeds 正常路径：每次都带上一次拿到的 version，
// 连续 5 次 update 必须次次成功，版本逐格递增。
//
// ⭐ 它挡的不是"没人覆盖"，是"挂了指不对地方"。
//
// 我实测过：把 CAS 条件改成恒假（模拟写得过于严格、永远返回冲突），
// 现有的 11 个用例会红 —— 包括并发那条。所以覆盖是有的。
// 平台侧原本的预测是"并发测试照样全绿"，那条不成立，我验之前就接受了它。
//
// 但红的 11 条里有 8 条是 TestCommit_* —— 因为它们都拿 Update 当前置步骤，
// CAS 一坏，前置就崩，报错全落在 Commit 上。照着那份输出去查，
// 人会从 Commit 开始 debug，而 bug 在 Update。
//
// 这条用例的价值因此是【诊断精度】而不是覆盖率：它红的时候只说一句
// "第 N 次串行 update 应当成功，实际 VERSION_CONFLICT"，指向就是对的。
func TestCAS_SerialAdvanceAlwaysSucceeds(t *testing.T) {
	ctx, s, _ := newStore(t)
	id := uuid.NewString()
	mustDraft(t, ctx, s, id, "购物车结算")

	version := 0
	for i := 1; i <= 5; i++ {
		r := s.Update(ctx, id, version, completeDraft("购物车结算", 2))
		if r.Code != model.OK {
			t.Fatalf("第 %d 次串行 update 应当成功，实际 %s: %s", i, r.Code, r.Message)
		}
		if r.Row.Version != i {
			t.Fatalf("第 %d 次 update 后 version 应为 %d，实际 %d", i, i, r.Row.Version)
		}
		version = r.Row.Version
	}

	// 收尾：正常路径推进完，commit 仍然走得通 —— 否则"能改但提交不了"也是死路
	if r := s.Commit(ctx, id, version); r.Code != model.OK {
		t.Fatalf("串行推进后 commit 应当成功，实际 %s: %s", r.Code, r.Message)
	}
}
